yarp/yarp_compiler.c

```diff
@@ -27,19 +39,79 @@ parse_number(const yp_node_t *node) {
     memcpy(buffer, start, length);
 
     buffer[length] = '\0';
-    VALUE number = rb_cstr_to_inum(buffer, -10, Qfalse);
+    VALUE number = DBL2NUM(rb_cstr_to_dbl(buffer, 0));
 
     free(buffer);
     return number;
 }
 
+static VALUE
+parse_rational(const yp_node_t *node)
+{
+    const uint8_t *start = node->location.start;
+    const uint8_t *end = node->location.end - 1;
+    size_t length = end - start;
+
+    char *buffer = malloc(length + 1);
+    memcpy(buffer, start, length);
+
+    buffer[length] = '\0';
+
+    VALUE res;
+    if (YP_NODE_TYPE_P(((yp_rational_node_t *)node)->numeric, YP_NODE_FLOAT_NODE)) {
+        char *decimal = memchr(buffer, '.', length);
+        RUBY_ASSERT(decimal);
+        size_t seen_decimal = decimal - buffer;
+        size_t fraclen = length - seen_decimal - 1;
+        memmove(decimal, decimal + 1, fraclen + 1);
+
+        VALUE v = rb_cstr_to_inum(buffer, 10, false);
+
+        res = rb_rational_new(v, rb_int_positive_pow(10, fraclen));
+    }
+    else {
+        RUBY_ASSERT(YP_NODE_TYPE_P(((yp_rational_node_t *)node)->numeric, YP_NODE_INTEGER_NODE));
+        VALUE number = rb_cstr_to_inum(buffer, 10, false);
```

Comment:
Now that we know we can parse integers without allocating, you could use the same technique we use in `parse_integer` here to avoid allocating out into the buffer.